accesstoken.go

```diff
@@ -34,34 +35,46 @@ func (c *WorkwxApp) syncAccessToken() error {
 	return nil
 }
 
-func (c *WorkwxApp) accessTokenRefresher() {
+func (c *WorkwxApp) accessTokenRefresher(ctx context.Context) {
 	const refreshTimeWindow = 30 * time.Minute
 	const minRefreshDuration = 5 * time.Second
 
-	// TODO: context cancellation
+	var waitDuration time.Duration = 0
 	for {
-		retryer := backoff.NewExponentialBackOff()
-		err := backoff.Retry(c.syncAccessToken, retryer)
-		if err != nil {
-			// wtf
-			// TODO: logging
-			_ = err
-		}
+		select {
+		case <-time.After(waitDuration):
+			retryer := backoff.NewExponentialBackOff()
+			err := backoff.Retry(c.syncAccessToken, retryer)
+			if err != nil {
+				// wtf
+				// TODO: logging
+				_ = err
+			}
 
-		waitUntilTime := c.lastRefresh.Add(c.tokenExpiresIn).Add(-refreshTimeWindow)
-		waitDuration := time.Until(waitUntilTime)
+			waitUntilTime := c.lastRefresh.Add(c.tokenExpiresIn).Add(-refreshTimeWindow)
+			waitDuration = time.Until(waitUntilTime)
 
-		if waitDuration < minRefreshDuration {
-			waitDuration = minRefreshDuration
+			if waitDuration < minRefreshDuration {
+				waitDuration = minRefreshDuration
+			}
+		case <-ctx.Done():
```

Comment:
This is not enough, because `backoff.Retry` is blocking, meaning you can't interrupt a series of retrying syncs with this `select` alone.

I consulted the docs and found [this](https://pkg.go.dev/github.com/cenkalti/backoff/v4?tab=doc#BackOffContext), maybe you should make use of it?

Reply:
Yeah, you're right. I missed it. And I have fixed it.